### PR TITLE
CES-9802	- Configure NICs' PXE boot settings for UEFI

### PR DESCRIPTION
**Pull request by digambar15**

### CODE REVIEW

File: src/pilot/config_idrac.py [hunk 1]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
         drac_client,
         ip_service_tag,
         pxe_nic_id,
-        node):
+        node, target_boot_mode):
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Sure

File: src/pilot/config_idrac.py [hunk 2]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
     LOG.info("Configuring NIC {} on {} to PXE boot".format(
         pxe_nic_id, ip_service_tag))
 
+    if target_boot_mode == boot_mode_helper.DRAC_BOOT_MODE_UEFI:
+        return configure_uefi_boot_mode_nics(drac_client,
+                                             pxe_nic_id)
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: okay

File: src/pilot/config_idrac.py [hunk 3]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
+    else:
+        return configure_bios_boot_mode_nics(drac_client,
+                                             ip_service_tag, pxe_nic_id)
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: okay

File: src/pilot/config_idrac.py [hunk 4]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
+    settings = {
+        "PxeDev1EnDis": "Enabled",
+        "PxeDev2EnDis": "Disabled",
+        "PxeDev3EnDis": "Disabled",
+        "PxeDev4EnDis": "Disabled",
+        "PxeDev1Interface": pxe_nic_id,
+        "PxeDev1Protocol": "IPv4",
+        "PxeDev1VlanEnDis": "Disabled"
+    }
     job_ids = []
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py [hunk 5]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
+
+
+def configure_bios_boot_mode_nics(drac_client, ip_service_tag, pxe_nic_id):
+    job_ids = []
+    reboot_required = False
+    provisioning_mac = None
     for nic in drac_client.list_nics(sort=True):
         result = None
         nic_id = nic.id
-
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py [hunk 6]
```diff
@@ -138,7 +176,7 @@ def configure_nics_boot_settings(
             except NotFound:
                 LOG.warn("Unable to check the legacy boot protocol of NIC {} "
                          "on {}, and so cannot set it to None".format(
-                             nic_id, ip_service_tag))
+                          nic_id, ip_service_tag))
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Its resored

File: src/pilot/config_idrac.py [hunk 7]
```diff
@@ -371,7 +409,7 @@ def config_idrac(instack_lock,
         configure_nics_boot_settings(drac_client,
                                      ip_service_tag,
                                      pxe_nic_fqdd,
-                                     node)
+                                     node, target_boot_mode)
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: ok

File: src/pilot/config_idrac.py [hunk 8]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
 
+    if target_boot_mode == boot_mode_helper.DRAC_BOOT_MODE_UEFI:
+        return configure_uefi_boot_mode_nics(drac_client,
+                                             pxe_nic_id)
+    else:
+        return configure_bios_boot_mode_nics(drac_client,
+                                             ip_service_tag, pxe_nic_id)
+
+
+def configure_uefi_boot_mode_nics(drac_client, pxe_nic_id):
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: okay

File: src/pilot/config_idrac.py [hunk 9]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
+        job_id = drac_client.commit_pending_bios_changes(reboot=False,
+                                                         start_time=None)
+        job_ids.append(job_id)
 
+    # Note that "commit_required" is actually "reboot_required" under the
+    # covers
+    return commit_required, job_ids, provisioning_mac
+
+
+def configure_bios_boot_mode_nics(drac_client, ip_service_tag, pxe_nic_id):
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: ok

File: src/pilot/config_idrac.py [hunk 10]
```diff
@@ -110,18 +110,56 @@ def configure_nics_boot_settings(
...
+def configure_uefi_boot_mode_nics(drac_client, pxe_nic_id):
+    settings = {
+        "PxeDev1EnDis": "Enabled",
+        "PxeDev2EnDis": "Disabled",
+        "PxeDev3EnDis": "Disabled",
+        "PxeDev4EnDis": "Disabled",
+        "PxeDev1Interface": pxe_nic_id,
+        "PxeDev1Protocol": "IPv4",
+        "PxeDev1VlanEnDis": "Disabled"
+    }
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Please reduce the scope of `settings`. It's only used inside the `if` statement below, so define it in its body.

**Comment by rpioso** [comment]
It appears you overlooked several comments in my review. Recall that GitHub hides comments beyond a certain number. You must click on `Load more ...` to display them. Those hidden comments include the more substantive ones.

Also, I added a comment. It's immediately above and involves the scope of the `settings` variable.

Finally, please push the resolution to my review and comment as a new and separate commit on the `configuring_nics_pxe_boot` branch. **Do not** forcibly overwrite the existing commit.

**Comment by digambar15** [comment]
Ok, I am merging this PR

**Pull request merged**